Source/WTF/wtf/StdLibExtras.h

```diff
@@ -764,14 +764,15 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
 template<typename T, typename U, std::size_t Extent>
 constexpr std::span<T, Extent == std::dynamic_extent ? std::dynamic_extent : (sizeof(U) * Extent) / sizeof(T)> spanReinterpretCast(std::span<U, Extent> span)
 {
+    static_assert(std::is_const_v<T> || (!std::is_const_v<T> && !std::is_const_v<U>), "spanReinterpretCast will not remove constness from source");
+    static_assert(!std::is_same_v<std::remove_const_t<T>, std::remove_const_t<U>>, "Unnecessary call to spanReinterpretCast");
```

Comment:
Personally, I do not like this. This makes span adoption harder because we have to write *more* code in templated functions like in Decoder.h to do the same thing (it's not just in Decoder.h, I'm hitting the exact same issue in another patch).

I'd be OK with that still if there were clear benefits. However, it is not clear to make the benefits are worth it here. There is no cost to an unnecessary reinterpret cast, is there?

Reply:
My opinion only:

If this makes template code annoying to write a second time then I agree it could be removed. In templates it can be useful to be able to cast without being so careful about the "null case" where the types match.

But in general outside a template type situation a mysterious cast isn't good, so one that says `reinterpret_cast` but isn't really doing anything at all is a small negative thing, not because it does harm but because it can cause slight confusion.

So if the bad template thing happens only **once**, I'd slightly lean toward keeping it.

Reply:
As I've mentioned, it's not once anymore. The reason I'm commenting is because I've just hit this during span adoption.

Reply:
```
In file included from /Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/persistence/PersistentCoders.cpp:26:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WTF/config.h:31:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/FastMalloc.h:26:
/Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/StdLibExtras.h:768:19: error: static assertion failed due to requirement '!std::is_same_v<unsigned char, unsigned char>': Unnecessary call to spanReinterpretCast
  768 |     static_assert(!std::is_same_v<std::remove_const_t<T>, std::remove_const_t<U>>, "Unnecessary call to spanReinterpretCast");
/Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/persistence/PersistentCoders.cpp:119:40: note: in instantiation of function template specialization 'WTF::spanReinterpretCast<unsigned char, unsigned char, 18446744073709551615UL>' requested here
  119 |     if (!decoder.decodeFixedLengthData(spanReinterpretCast<uint8_t>(buffer)))
/Volumes/Work/WebKit/OpenSource/Source/WTF/wtf/persistence/PersistentCoders.cpp:143:16: note: in instantiation of function template specialization 'WTF::Persistence::decodeStringText<unsigned char>' requested here
  143 |         return decodeStringText<LChar>(decoder, *length);
1 error generated.
```

Reply:
> But in general outside a template type situation a mysterious cast isn't good, so one that says reinterpret_cast but isn't really doing anything at all is a small negative thing, not because it does harm but because it can cause slight confusion.

I agree that unnecessary casts aren't good. However, I don't know that this is a problem we currently need to solve. We haven't found any cast of unnecessary span casting outside template code yet. However, we've found 2 instances where we need to write more code to do achieve the same behavior, which is unfortunate.

As we're doing more and more span adoption for security reasons, I don't think we want to make that work more painful than it needs to be.

Reply:
I've hit this a third time in decodeStringText() in DaemonCoders.h.

Reply:
You can see those cases in https://github.com/WebKit/WebKit/pull/35793. For now, I'm proposing to revert this new check (which is what I'm currently doing as part of my PR). I can be convinced to go the hard way and add constexpr is_same_v checks if people really prefer that but it's not my preference.

Reply:
Sure, I agree, let’s take it this check out. It was my suggestion and you have convinced me!